File: Functions/computeCumulativeCharge.m
function cumulative_integral = computeCumulativeCharge(timeS, current)
% computeCumulativeCharge - Calculate cumulative charge integral
%
% This function calculates the cumulative charge (capacity) by integrating
% current over time. It processes each continuous segment separately,
% splitting by NaN gaps to maintain data integrity.
%
% Inputs:
%   timeS   - Time array in seconds (may contain NaN values to mark gaps)
%   current - Current array in Amperes (corresponding to timeS)
%
% Outputs:
%   cumulative_integral - Cumulative charge in Ampere-seconds (As)
%                         Contains NaN values at gap positions

fprintf('Computing cumulative charge integral...\n');
tic;  % Restart timer

cumulative_integral = [];
start_idx = 1;

% Loop through the time signal to find segments
for i = 2:length(timeS)
    if isnan(timeS(i))
        % Process the segment before this NaN

        current_segment = current(start_idx:i-1);      
        % Compute the integral using trapezoidal rule
        if start_idx < i-1
            segment_integral = cumtrapz(timeS(start_idx:i-1), current_segment);
        else
            segment_integral = 0;
        end
        
%         Append the segment integral with NaN separator
        cumulative_integral = [cumulative_integral; NaN; segment_integral];

        
        % Reset the start index for the next segment
        start_idx = i+1;
    end
end

% Handle the last segment (after final NaN or if no NaN exists)
if start_idx==length(current)
    cumulative_integral = [cumulative_integral; 0];
else
    current_segment = current(start_idx:end);
    segment_integral = cumtrapz(timeS(start_idx:end), current_segment);
    cumulative_integral = [cumulative_integral; segment_integral];
end
fprintf('Cumulative charge calculation complete. (Elapsed: %.2f s)\n', toc);

end

File: Functions/computeCumulativeCharge.py
"""
computeCumulativeCharge - Calculate cumulative charge integral

This function calculates the cumulative charge (capacity) by integrating
current over time. It processes each continuous segment separately,
splitting by NaN gaps to maintain data integrity.
"""

import numpy as np
from scipy.integrate import cumtrapz
import time


def compute_cumulative_charge(time_s, current):
    """
    Calculate cumulative charge integral.
    
    Parameters
    ----------
    time_s : numpy.ndarray
        Time array in seconds (may contain NaN values to mark gaps)
    current : numpy.ndarray
        Current array in Amperes (corresponding to time_s)
    
    Returns
    -------
    cumulative_integral : numpy.ndarray
        Cumulative charge in Ampere-seconds (As)
        Contains NaN values at gap positions
    """
    
    print('Computing cumulative charge integral...')
    start_time = time.time()
    
    # Pre-allocate output array
    cumulative_integral = np.full(len(time_s), np.nan)
    
    # Find NaN positions
    nan_mask = np.isnan(time_s)
    
    # Find continuous segments (between NaNs)
    # Add sentinels at start and end
    boundaries = np.concatenate([[True], nan_mask, [True]])
    # Find transitions
    diff = np.diff(boundaries.astype(int))
    
    # Start indices are where we go from NaN (or start) to valid data
    start_indices = np.where(diff == -1)[0]
    # End indices are where we go from valid data to NaN (or end)
    end_indices = np.where(diff == 1)[0]
    
    # Process each segment
    for start_idx, end_idx in zip(start_indices, end_indices):
        if end_idx > start_idx:  # Valid segment
            segment_time = time_s[start_idx:end_idx]
            segment_current = current[start_idx:end_idx]
            
            # Use scipy's cumtrapz for efficient cumulative integration
            segment_cumulative = cumtrapz(segment_current, segment_time, initial=0)
            
            # Store in output array
            cumulative_integral[start_idx:end_idx] = segment_cumulative
    
    elapsed = time.time() - start_time
    print(f'Cumulative charge calculation complete. (Elapsed: {elapsed:.2f} s)')
    
    return cumulative_integral
